File: 380-insert-delete-getrandom-o1.java
class RandomizedSet {
    List<Integer> nums;
    Map<Integer, Integer> numIndexMap;
    Random random;
    public RandomizedSet() {
        nums = new ArrayList<>();
        numIndexMap = new HashMap<>();
        random = new Random();
    }
    
    public boolean insert(int val) {
        if(numIndexMap.containsKey(val)) {
            return false;
        }
        
        numIndexMap.put(val, nums.size());
        nums.add(val);
        return true;
    }
    
    public boolean remove(int val) {
        if(!numIndexMap.containsKey(val)) {
            return false;
        }
        
        // Overwrite nums's element that equals to target value with last element
        int index = numIndexMap.get(val);
        int lastElement = nums.get(nums.size() - 1);
        nums.set(index, lastElement);    
        numIndexMap.put(lastElement, index);
        
        // Remove last element
        numIndexMap.remove(val);
        nums.remove(nums.size() - 1);
        return true;
    }
    
    public int getRandom() {
        int randomIndex = random.nextInt(nums.size());
        return nums.get(randomIndex);
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: 380_Insert_Delete_GetRandom_O1.py
I use a list called values to store all elements, and a hash map index_map that maps each value to its index in the list.
Insert: append the element and record its index.
Remove: swap the element with the last element, update the map, then pop it.
GetRandom: choose a random element from values.
This makes insert, remove, and getRandom all O(1).

import random

class RandomizedSet:

    def __init__(self):
        self.values = []                   # list of stored elements
        self.index_map = {}                # value -> index in values

    def insert(self, value: int) -> bool:
        if value in self.index_map:
            return False

        self.index_map[value] = len(self.values)
        self.values.append(value)
        return True

    def remove(self, value: int) -> bool:
        if value not in self.index_map:
            return False

        remove_index = self.index_map[value]
        last_value = self.values[-1]

        # Move last element into the removed spot
        self.values[remove_index] = last_value
        self.index_map[last_value] = remove_index

        # Remove the old entry
        self.values.pop()
        del self.index_map[value]

        return True

    def getRandom(self) -> int:
        return random.choice(self.values)
